systemtests/src/main/java/io/enmasse/systemtest/messagingclients/MessagingClientRunner.java

```diff
@@ -0,0 +1,111 @@
+    public List<ExternalMessagingClient> sendAndReceive(MessagingEndpoint endpoint, boolean waitReceivers, Map<ClientArgument, Object> extraSenderArgs,
+                           Map<ClientArgument, Object> extraReceiverArgs, String senderAddress, String ... receiverAddresses) throws Exception {
+        int expectedMsgCount = 10;
+
+        ExecutorService executor = Executors.newFixedThreadPool(1 + receiverAddresses.length);
+        List<ExternalMessagingClient> clients = new ArrayList<>();
+        Endpoint e = new Endpoint(endpoint.getStatus().getHost(), getPort("AMQP", endpoint));
+        ExternalMessagingClient senderClient = new ExternalMessagingClient(false)
+                .withClientEngine(new RheaClientSender())
+                .withMessagingRoute(e)
+                .withAddress(senderAddress)
+                .withCount(expectedMsgCount)
+                .withMessageBody("msg no. %d")
+                .withAdditionalArgument(ClientArgument.CONN_AUTH_MECHANISM, "ANONYMOUS")
+                .withTimeout(60);
+
+        if (extraSenderArgs != null) {
+            for (Map.Entry<ClientArgument, Object> arg : extraSenderArgs.entrySet()) {
+                senderClient.withAdditionalArgument(arg.getKey(), arg.getValue());
+            }
+        }
+
+        for (String receiverAddress : receiverAddresses) {
+            ExternalMessagingClient receiverClient = new ExternalMessagingClient(false)
+                    .withClientEngine(new RheaClientReceiver())
+                    .withMessagingRoute(e)
+                    .withAddress(receiverAddress)
+                    .withCount(expectedMsgCount)
+                    .withAdditionalArgument(ClientArgument.CONN_AUTH_MECHANISM, "ANONYMOUS")
+                    .withTimeout(60);
+
+            if (extraReceiverArgs != null) {
+                for (Map.Entry<ClientArgument, Object> arg : extraReceiverArgs.entrySet()) {
+                    receiverClient.withAdditionalArgument(arg.getKey(), arg.getValue());
+                }
+            }
+
+            clients.add(receiverClient);
+        }
+
+        List<Future<Boolean>> receiverResults = new ArrayList<>();
+        for (ExternalMessagingClient receiverClient : clients) {
+            receiverResults.add(executor.submit((Callable<Boolean>) receiverClient::run));
+        }
+
+        if (waitReceivers) {
+            // To ensure receivers are attached and ready
+            Thread.sleep(10_000);
+        }
+
+        Future<Boolean> senderResult = executor.submit((Callable<Boolean>) senderClient::run);
+        clients.add(senderClient);
+        assertTrue(senderResult.get(1, TimeUnit.MINUTES), "Sender failed, expected return code 0");
+        for (Future<Boolean> receiverResult : receiverResults) {
+            assertTrue(receiverResult.get(1, TimeUnit.MINUTES), "Receiver failed, expected return code 0");
+        }
+        return clients;
+    }
+
+
+    public int getPort(String protocol, MessagingEndpoint endpoint) {
```

Comment:
I think this method is more common and it should be in MessagingEndpointResource class as a static

Reply:
I've moved this to MessagingAdressResouce as static.